python/pyiceberg/catalog/dynamodb.py

```diff
@@ -0,0 +1,733 @@
+                    },
+                },
+                ConditionExpression=f"attribute_exists({DYNAMODB_COL_NAMESPACE})",
+            )
+        except self.dynamodb.exceptions.ConditionalCheckFailedException as e:
+            raise NoSuchNamespaceError(f"Database does not exist: {database_name}") from e
+        except (
+            self.dynamodb.exceptions.ProvisionedThroughputExceededException,
+            self.dynamodb.exceptions.ResourceNotFoundException,
+            self.dynamodb.exceptions.ItemCollectionSizeLimitExceededException,
+            self.dynamodb.exceptions.TransactionConflictException,
+            self.dynamodb.exceptions.RequestLimitExceeded,
+            self.dynamodb.exceptions.InternalServerError,
+        ) as e:
+            raise GenericDynamoDbError(e.message) from e
+
+    def list_tables(self, namespace: Union[str, Identifier]) -> List[Identifier]:
+        """List tables under the given namespace in the catalog (including non-Iceberg tables)
+
+        Args:
+            namespace (str | Identifier): Namespace identifier to search.
+
+        Returns:
+            List[Identifier]: list of table identifiers.
+        """
+
+        database_name = self.identifier_to_database(namespace, NoSuchNamespaceError)
+
+        paginator = self.dynamodb.get_paginator("query")
+
+        try:
+            page_iterator = paginator.paginate(
+                TableName=ICEBERG,
+                IndexName=DYNAMODB_NAMESPACE_GSI,
+                KeyConditionExpression=f"{DYNAMODB_COL_NAMESPACE} = :namespace ",
+                ExpressionAttributeValues={
+                    ":namespace": {
+                        "S": database_name,
+                    }
+                },
+            )
+        except (
+            self.dynamodb.exceptions.ProvisionedThroughputExceededException,
+            self.dynamodb.exceptions.ResourceNotFoundException,
+            self.dynamodb.exceptions.RequestLimitExceeded,
+            self.dynamodb.exceptions.InternalServerError,
+        ) as e:
+            raise GenericDynamoDbError(e.message) from e
+
+        table_identifiers = []
+        for page in page_iterator:
+            for item in page["Items"]:
+                _dict = _convert_dynamo_item_to_regular_dict(item)
+                identifier_col = _dict[DYNAMODB_COL_IDENTIFIER]
+                if identifier_col == DYNAMODB_NAMESPACE:
+                    continue
+
+                table_identifiers.append(self.identifier_to_tuple(identifier_col))
+
+        return table_identifiers
+
+    def list_namespaces(self, namespace: Union[str, Identifier] = ()) -> List[Identifier]:
+        """
+        List top-level namespaces from the catalog.
+        We do not support hierarchical namespace.
+
+        Returns:
+            List[Identifier]: a List of namespace identifiers
+        """
+
+        if namespace:
+            raise ValidationError("This API is not supported for hierarchical namespaces.")
```

Comment:
This is a good point. 

I actually didn't think just returning an empty list would be a good idea. Because in that case if a user sends in a namespace they will just get an empty list. That will imply their dataset does not have any database under their provided namespace which which may not be the truth and without getting any exception from the app saying that this feature is not supported and it's not like the answer to your question is an empty list. 

I like to hear @jackye1995 and @Fokko option on this too and decide whether we want to go with an empty list or throw.

Reply:
Discussion here:
https://github.com/apache/iceberg/pull/6646#discussion_r1097947897